euphonic/writers/phonon_website.py

```diff
@@ -0,0 +1,218 @@
+        atom_types=crystal.atom_type.tolist(),
+        atom_numbers=list(map(get_z, crystal.atom_type)),
+        formula=symbols_to_formula(crystal.atom_type),
+        atom_pos_red=crystal.atom_r.tolist(),
+        atom_pos_car=(crystal.atom_r @ crystal.cell_vectors
+                      ).to("angstrom").magnitude.tolist()
+    )
+
+
+def _remove_breaks(distances: np.ndarray, btol: float = 10.) -> list[int]:
+    """Collapse large breaks in cumulative-distance array
+
+    These correspond to discontinuous regions of the x-axis: in euphonic
+    plotting this is usually handled by splitting the spectrum and plotting
+    to new axes, but Phonon Website does not handle this.
+
+    Data is modified in-place
+
+    A list of identified breakpoints is returned
+
+    """
+    diff = np.diff(distances)
+    median = np.median(diff)
+    breakpoints = np.where((diff / median) > btol)[0] + 1
+
+    for breakpoint in reversed(breakpoints):
+        distances[breakpoint:] -= (distances[breakpoint]
+                                   - distances[breakpoint - 1])
+
+    return breakpoints.tolist()
+
+
+def _find_duplicates(distances: np.ndarray) -> list[int]:
+    """Identify breakpoints where a q-point is repeated in list
+
+    Return the higher index of two points: this works more nicely with phonon
+    website interpretation of line break points.
+    """
+
+    diff = np.diff(distances)
+    duplicates = np.where(diff == 0.)[0] + 1
+
+    return duplicates.tolist()
+
+
+def _combine_neighbouring_labels(x_tick_labels: XTickLabels) -> XTickLabels:
+    """Merge neighbouring labels in x_tick_label list
+
+    If labels are the same, only keep one.
+
+    If labels are different, join with |
+
+    e.g.::
+
+      [(1, "X"), (2, "X"), (4, "A"), (7, "Y"), (8, "Z")]
+
+      -->
+
+      [(1, "X"), (4, "A"), (7, "Y|Z")]
+
+    """
+    labels = dict(x_tick_labels)
+
+    for (index, next_index) in pairwise(sorted(labels)):
+
+        if index + 1 == next_index:  # Labels are neighbours
+
+            if labels[index] != labels[next_index]:  # Combine differing labels
+                labels[next_index] = f"{labels[index]}|{labels[next_index]}"
+
+            del labels[index]  # Drop redundant label
+
+    return sorted(labels.items())
+
+
+def _modes_to_phonon_website_dict(modes: QpointPhononModes,
+                                  name: str | None = None,
+                                  repetitions: tuple[int, int, int] = (2, 2, 2),
+                                  x_tick_labels: XTickLabels | None = None,
+                                  ) -> PhononWebsiteData:
+    qpts = modes.qpts
+    eigenvectors = modes.eigenvectors
+
+    abscissa = _calc_abscissa(modes.crystal.reciprocal_cell(), qpts)
+
+    duplicates = _find_duplicates(abscissa)
+    breakpoints = _remove_breaks(abscissa)
+
+    breakpoints = sorted(set([0] + duplicates + breakpoints + [len(abscissa)]))
+    line_breaks = [(start, end) for start, end in pairwise(breakpoints)]
```

Comment:
Is this not `list(pairwise(breakpoints))`?

Reply:
Yes, some numerical offsets have been changed/removed over time 😆